Fixing bug where native.exe --<tab> would not call native completer

powercode opened the pull request:
A case was missing when parsing the processing the current token. The result was that the registered argument completer for a native command was not called if the wordToComplete was '--'.

Commit:
Extract variable tokenAtCursorText

from tokenAtCursor.Text.

Commit:
Adding support for double-dash native completion

Commit:
Cleanup

coaleasing null checks and type patterns.

Commit:
Adding failing test

iSazonov commented on src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:
```diff
@@ -156,7 +156,7 @@ private static bool CompleteAgainstSwitchFile(Ast lastAst, Token tokenBeforeCurs
 
             var errorStatement = lastAst as ErrorStatementAst;
-            if (errorStatement != null && errorStatement.Flags != null && errorStatement.Kind != null && tokenBeforeCursor != null &&
+            if (errorStatement?.Flags != null && errorStatement.Kind != null && tokenBeforeCursor != null &&
```
The same about c# 7.0 pattern.

iSazonov commented on src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:
```diff
@@ -173,7 +173,7 @@ private static bool CompleteAgainstSwitchFile(Ast lastAst, Token tokenBeforeCurs
 
                 errorStatement = pipeline.Parent as ErrorStatementAst;
-                if (errorStatement == null || errorStatement.Kind == null || errorStatement.Flags == null)
+                if (errorStatement?.Kind == null || errorStatement.Flags == null)
```
The same about c# 7.0 pattern.

iSazonov commented on src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:
```diff
@@ -206,15 +206,13 @@ private static bool CompleteAgainstStatementFlags(Ast scriptAst, Ast lastAst, To
             // Handle "switch -f<tab>"
             var errorStatement = lastAst as ErrorStatementAst;
-            if (errorStatement != null && errorStatement.Kind != null)
+            if (errorStatement?.Kind != null)
```
The same about c# 7.0 pattern.

iSazonov commented on src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:
```diff
@@ -882,7 +871,7 @@ internal static TypeName FindTypeNameToComplete(ITypeName type, IScriptPosition
             var arrayTypeName = type as ArrayTypeName;
             if (arrayTypeName != null)
```
The same about c# 7.0 pattern.

iSazonov replied:
Closed.

iSazonov commented:
@powercode Please add description in the PR. This can be useful for documentation.

Commit:
Fixing styling issues

Addressing review comments by iSazonov

powercode commented:
@iSazonov Thanks for the review. 
I've addressed your comments, except for the one about the registration of TabCompleters.
These are stored in private variables in the ExecutionContext, and as far as I know, there is no way to Unregister-ArgumentCompleter. @lzybkr ?

iSazonov commented:
I think it's even better for tests to register several TabCompleters in BeforeAll and not remove them.

iSazonov commented on test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:
```diff
@@ -0,0 +1,72 @@
+                else {
+                    return "unexpected wordtocomplete"
+                }
+
```
Please remove extra line.

iSazonov replied:
Closed.

iSazonov commented on test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:
```diff
@@ -0,0 +1,72 @@
+                else {
+                    return "unexpected wordtocomplete"
+                }
+
```
Please remove extra line.

iSazonov replied:
Closed.

iSazonov commented on test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:
```diff
@@ -0,0 +1,72 @@
+Describe TabCompletion -Tags CI {
```
Closed.

iSazonov commented on test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:
```diff
@@ -0,0 +1,72 @@
+        BeforeAll {
+            $nativeCommand = (Get-Command -CommandType Application -TotalCount 1).Name
+        }
+        It 'Completes native commands with -' {         
```
Closed.

iSazonov commented on test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:
```diff
@@ -0,0 +1,72 @@
+                    return "-flag"
+                }
+                else {
+                    return 'unexpected wordtocomplete'
```
Closed.

iSazonov commented on test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:
```diff
@@ -0,0 +1,72 @@
+            }
+            $line = "$nativeCommand -"
+            $res = TabExpansion2 -inputScript $line -cursorColumn $line.Length
+            $res.CompletionMatches.Count | Should be 1
```
Closed.

iSazonov commented on test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:
```diff
@@ -0,0 +1,72 @@
+                else {
+                    return 'unexpected wordtocomplete'
+                }
+
```
Closed.

iSazonov commented on test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:
```diff
@@ -0,0 +1,72 @@
+                else {
+                    return 'unexpected wordtocomplete'
+                }
+
```
Closed.

iSazonov commented on test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:
```diff
@@ -0,0 +1,72 @@
+            $res.CompletionMatches.CompletionText | Should be "-option"                        
+        }
+    }
+}
```
Closed.

Commit:
manual merge of TabCompletion.Tests.ps1

powercode commented:
Rebased master and fixed conflicting test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1

lzybkr committed:
Merge branch 'master' into native-arg-hyphen

lzybkr reviewed (commented):
Would it be unreasonable to ask for logic code changes only?

The C# 7 language usage is nice, but I'd rather see those sorts of changes in a distinct PR.

Commit:
Removing usage of C#7 constructs

iSazonov commented:
Sometimes mantainers aren't predictable - I just asked what I was getting myself.

lzybkr commented:
Sometimes people aren't predictable.

PowerShell is no longer my full time job - so when I ask for something like this, it's to make things easier for me. It's already somewhat hard to keep up as a maintainer with an unrelated full time job and non-computer related responsibilities and hobbies.

If you could see our pre-GitHub history, you would certainly see that I've been guilty of changes like this, but I blame not having a good enough VCS to make it easier. So I have plenty of personal experience with the difficulties of looking at changes that mix style and logical changes - and I do think a cleaner and easier to understand history is valuable in the long run, even if it wasn't painful to review something like this in the short term.

iSazonov commented:
I meant that often we contribute and discuss _before_ the appointment of mantainer and before his posts - so we can _not predict_ a review style and facilitate the mantainer's work.
I believe our workflow process could be improved, I guess we could take more from the CoreCLR/CoreFX workflows.
/cc @joeyaiello

lzybkr commented:
Note that CoreFX says this:

```
DO NOT send PRs for upgrading code to use newer language features, though it's ok to use newer language features as part of new code that's written.
```

See [here](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/contributing.md).

iSazonov commented:
Yes, they were more clearly writing the workflow. We have to do the same.

lzybkr commented:
:shipit:

lzybkr commented on test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:
```diff
@@ -101,6 +100,75 @@ Describe "TabCompletion" -Tags CI {
+
+    Context NativeCommand {
+        BeforeAll {
+            $nativeCommand = (Get-Command -CommandType Application -TotalCount 1).Name
```
It might be wiser to create an external command on testdrive and change the path, but that's a tiny bit annoying to make it portable, so I'll merge as is.
